castai/client.go

```diff
@@ -33,17 +36,54 @@ func NewClient(log *logrus.Logger, rest *resty.Client, clusterID string) Client
 }
 
 // NewDefaultClient configures a default instance of the resty.Client used to do HTTP requests.
-func NewDefaultClient(url, key string, level logrus.Level, binVersion *config.ClusterControllerVersion) *resty.Client {
-	client := resty.New()
+func NewDefaultClient(url, key string, level logrus.Level, binVersion *config.ClusterControllerVersion, defaultTimeout time.Duration) (*resty.Client, error) {
+	clientTransport, err := createHTTPTransport()
+	if err != nil {
+		return nil, err
+	}
+
+	client := resty.NewWithClient(&http.Client{
+		Timeout:   defaultTimeout,
+		Transport: clientTransport,
+	})
+
 	client.SetHostURL(url)
-	client.SetTimeout(5 * time.Minute) // Hard timeout for any request.
 	client.Header.Set(headerAPIKey, key)
 	client.Header.Set(headerUserAgent, "castai-cluster-controller/"+binVersion.Version)
 	if level == logrus.TraceLevel {
 		client.SetDebug(true)
 	}
 
-	return client
+	return client, nil
+}
+
+func createHTTPTransport() (*http.Transport, error) {
+	t1 := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout: 15 * time.Second,
+		}).DialContext,
+		ForceAttemptHTTP2:     true,
+		MaxIdleConns:          100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   5 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+	}
+
+	t2, err := http2.ConfigureTransports(t1)
```

Comment:
We are not using http2 in controller. Could you elaborate why http2?

Reply:
we are, it is default from resty library

Reply:
I could not find Resty is calling `http2.ConfigureTransports`. Moreover to my understanding Resty relies on `http` (even though it sets `ForceAttemptHTTP2` to true) package: https://github.com/go-resty/resty/blob/v2/go.mod#L6